hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/SCMNodeManager.java

```diff
@@ -240,8 +247,19 @@ public VersionResponse getVersion(SCMVersionRequestProto versionRequest) {
   @Override
   public RegisteredCommand register(
       DatanodeDetails datanodeDetails, NodeReportProto nodeReport,
-      PipelineReportsProto pipelineReportsProto) {
-
+      PipelineReportsProto pipelineReportsProto,
+      LayoutVersionProto layoutInfo) {
+
+    if (layoutInfo != null) {
```

Comment:
Can we move this check to org.apache.hadoop.hdds.scm.server.SCMDatanodeProtocolServer#register? That way, a lot of test changes in this patch to incorporate the 4th parameter to NodeManager.register() can be avoided.

Reply:
I thought about it while I wrote this. I wanted all registration related processing for all the passed parameters to be consolidated in one place and it seems like NodeManager.register is the right place even though we have to add another parameter in bunch of places. Majority of these places are unit tests.

Reply:
Have you considered Optional<T> semantics here? Does it make code more readable?

Reply:
I would still need to check for condition where the argument is not valid.